src/vcf/record.cpp

```diff
@@ -518,97 +537,70 @@ namespace ebi
-                        message = tag_field + " " + field + " value is not a natural number, negatives not allowed"; 
-                        throw new Error;
-                    }
-                } else if (type == "Float") {
-                    // ...try to cast to float
-                    try {
-                        std::stof(value);
-                    } catch (std::out_of_range) {
-                        // It maybe a subnormal number
-                        std::stold(value);
-                    }
-                } else if (type == "Flag") {
-                    if (value.size() > 1) {
-                        if (status) {
-                            raise(std::make_shared<Error>(line, "There can be only 0 or 1 value"));
-                        } else {
-                            message = tag_field + " " + field + " value is not Flag, it can only have 0 or 1 value";
-                            throw new Error;
-                        }
-                    } else if (value.size() == 1) {
-                        int numeric_value = std::stoi(value);
-                        if (numeric_value != 0 && numeric_value != 1) {
-                            if (status) {
-                                raise(std::make_shared<Error>(line, "A flag must be 0 or 1"));
-                            } else {
-                                message = tag_field + " " + field + " value is not Flag, it must be (1/0/no value)";
-                                throw new Error;
-                            }
-                        }
-                    }
-                    // If no flag is provided then there is nothing to check
-                } else if (type == "Character") {
-                    // ...check the length is 1
-                    if (value.size() > 1) {
-                        raise(std::make_shared<Error>(line, "There can be only one character"));
-                    }
-                } else if (type == "String") {
-                    // ...do nothing, it is guaranteed it will be a string
-                }
-            } catch (...) {
-                if (status) {
-                    raise(std::make_shared<Error>(line, field + " does not match the meta specification Type=" + type));
-                } else {
-                    if (message.size() == 0) {
-                        message = tag_field + " " + field + " value is not " + type;
-                    }
-                    if (tag_field == "INFO") {
-                        throw new InfoBodyError{line, message, field};
-                    } else {
-                        throw new FormatBodyError{line, message};
-                    }
-                }
+                check_value_type(type, value, message);
+            } catch (std::exception &typeError) {
+                raise(std::make_shared<Error>(line, field + " does not match the meta specification Type=" + type + ". " + message));
```

Comment:
Ok, I think we can do a final refactor here. the purpose is to remove that message returned by reference to build the exception. I was thinking in putting a bit more detail in the inner functions like `check_value_type` and `check_field_cardinality`.

This is a good opportunity to address a suggestion a user made some time ago, that the messages would be more useful if they had (on top of the explanation of the error) the actual offending values (like "AC=2,3,4").

To guide the refactor, try to remove the `try-catch` in this function (and maybe `check_predefined_tag` if it's possible?), so at the end we have these `try`s and `catch`es:
- in `check_field_cardinality` and `check_value_type` throw std::invalid_argument with a message that include the explanation (like "Float provided instead of Integer") but also the exact offending value (`value`) and stating clearly  that it is a problem regarding the type or cardinality. Something like `value + " (float) was provided as value of Type=" + type` 
- in check_info,` catch(const std::invalid_argument &ex)`: throw an InfoBodyError with something like "INFO " + field.first + "=" + field.second + ": " + ex.what();
- in check_info, but for the predefined tags,` catch(const std::invalid_argument &ex)`: throw InfoBodyError with "INFO (defined in the specification) " + field.first + "=" + field.second + ": " + ex.what();
- in check_sample_subfields_cardinality_type, throw SamplesFieldBodyError similar message including the sample number